optuna/study/_optimize.py

```diff
@@ -217,111 +211,42 @@ def _run_trial(
         state = TrialState.FAIL
         func_err = e
         func_err_fail_exc_info = sys.exc_info()
-    else:
-        # TODO(hvy): Avoid checking the values both here and inside `Study.tell`.
-        values, values_conversion_failure_message = _check_and_convert_to_values(
-            len(study.directions), value_or_values, trial.number
-        )
-        if values_conversion_failure_message is not None:
-            state = TrialState.FAIL
-        else:
-            state = TrialState.COMPLETE
 
     if study._storage.is_heartbeat_enabled():
         assert stop_event is not None
         assert thread is not None
         stop_event.set()
         thread.join()
 
-    # `Study.tell` may raise during trial post-processing.
+    # `Study._tell` may raise during trial post-processing.
     try:
-        study.tell(trial, values=values, state=state)
+        # Note: call `Study._tell` instead of `Study.tell` to suppress warning.
+        frozen_trial = study._tell_with_warning(
+            trial, values=value_or_values, state=state, suppress_warning=True
+        )
     except Exception:
+        frozen_trial = study._storage.get_trial(trial._trial_id)
         raise
     finally:
-        if state == TrialState.COMPLETE:
-            study._log_completed_trial(trial, cast(List[float], values))
-        elif state == TrialState.PRUNED:
-            _logger.info("Trial {} pruned. {}".format(trial.number, str(func_err)))
-        elif state == TrialState.FAIL:
+        if frozen_trial.state == TrialState.COMPLETE:
+            study._log_completed_trial(frozen_trial)
+        elif frozen_trial.state == TrialState.PRUNED:
+            _logger.info("Trial {} pruned. {}".format(frozen_trial.number, str(func_err)))
+        elif frozen_trial.state == TrialState.FAIL:
             if func_err is not None:
-                _logger.warning(
-                    "Trial {} failed because of the following error: {}".format(
-                        trial.number, repr(func_err)
-                    ),
-                    exc_info=func_err_fail_exc_info,
-                )
-            elif values_conversion_failure_message is not None:
-                _logger.warning(values_conversion_failure_message)
-            else:
-                assert False, "Should not reach."
+                _log_failed_trial(frozen_trial, repr(func_err), exc_info=func_err_fail_exc_info)
+            elif "study_tell_warning" in frozen_trial.system_attrs:
+                _log_failed_trial(frozen_trial, frozen_trial.system_attrs["study_tell_warning"])
```

Comment:
Is an "else" meant to be logically possible here, i.e. a failed trial without any following warning log? If not, how about an `else: assert False, "Should not reach."` after these lines? On the other hand, if it's possible, a comment would be informative.

Reply:
It is not expected that a trial fails without func error or warning message (from Study.tell). I added "else" in a21c7be. Thank you for the advise!